core/src/main/java/jenkins/install/SetupWizard.java

```diff
@@ -0,0 +1,180 @@
+package jenkins.install;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.UUID;
+import java.util.logging.Logger;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+
+import hudson.BulkChange;
+import hudson.ExtensionList;
+import hudson.model.User;
+import hudson.model.UserProperty;
+import hudson.security.AuthorizationStrategy;
+import hudson.security.FullControlOnceLoggedInAuthorizationStrategy;
+import hudson.security.HudsonPrivateSecurityRealm;
+import hudson.security.PermissionAdder;
+import hudson.security.SecurityRealm;
+import hudson.security.csrf.DefaultCrumbIssuer;
+import hudson.util.PluginServletFilter;
+import jenkins.model.Jenkins;
+import jenkins.security.s2m.AdminWhitelistRule;
+
+/**
+ * A Jenkins instance used during first-run to provide a limited set of services while
+ * initial installation is in progress
+ */
+public class SetupWizard {
+    /**
+     * The security token parameter name
+     */
+    public static String initialSetupAdminUserName = "initial-setup-admin-user";
+
+    private final Logger LOGGER = Logger.getLogger(SetupWizard.class.getName());
+
+    public SetupWizard(Jenkins j) throws IOException {
+        User admin;
+        // Create an admin user by default with a 
+        // difficult password
+        if(j.getSecurityRealm() == null || j.getSecurityRealm() == SecurityRealm.NO_AUTHENTICATION) { // this seems very fragile
+            BulkChange bc = new BulkChange(j);
+            
+            HudsonPrivateSecurityRealm securityRealm = new HudsonPrivateSecurityRealm(false, false, null);
+            j.setSecurityRealm(securityRealm);
+            String randomUUID = UUID.randomUUID().toString().replace("-", "").toLowerCase(Locale.ENGLISH);
+            admin = securityRealm.createAccount(SetupWizard.initialSetupAdminUserName, randomUUID);
+            admin.addProperty(new SetupWizard.AuthenticationKey(randomUUID));
+
+            AuthorizationStrategy as = Jenkins.getInstance().getAuthorizationStrategy();
+            for (PermissionAdder adder : ExtensionList.lookup(PermissionAdder.class)) {
+                if (adder.add(as, admin, Jenkins.ADMINISTER)) {
+                    return;
```

Comment:
:bug: I think you were copying [this code](https://github.com/jenkinsci/jenkins/blob/12d96a567e88dc4b1e1fab18b81ba4a905773e9e/core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java#L298-L300) but out of context the `return` makes no sense.

Anyway this extension point is only useful insofar it might trigger an implementation such as [this](https://github.com/jenkinsci/matrix-auth-plugin/blob/d9565739e661fb58d3213f2a994cc3bc700035a3/src/main/java/hudson/security/GlobalMatrixAuthorizationStrategy.java#L381-L384), which should be impossible at this point since no plugins are loaded.